src/cascadia/TerminalSettingsModel/Profile.cpp

```diff
@@ -533,5 +539,141 @@ Json::Value Profile::ToJson() const
+                error << L"Self referencing keys in environment settings: ";
+                for (auto seenKey : seenKeys)
+                {
+                    error << L"'" << seenKey << L"', ";
+                }
+                error << L"'" << key << L"'";
+                throw winrt::hresult_error(WEB_E_INVALID_JSON_STRING, winrt::hstring(error.str()));
+            }
+            const std::wregex parameterRegex(L"\\$\\{env:(.*?)\\}");
+            auto& value = it->second;
+            auto textIter = value.rawValue.cbegin();
+            std::wsmatch regexMatch;
+            while (std::regex_search(textIter, value.rawValue.cend(), regexMatch, parameterRegex))
+            {
+                if (regexMatch.size() != 2)
+                {
+                    std::wstringstream error;
+                    error << L"Unexpected regex match count '" << regexMatch.size()
+                          << L"' (expected '2') when matching '" << std::wstring(textIter, value.rawValue.cend()) << L"'";
+                    throw winrt::hresult_error(WEB_E_INVALID_JSON_STRING, winrt::hstring(error.str()));
+                }
+                value.resolvedValue += std::wstring(textIter, regexMatch[0].first);
+                const auto referencedKey = regexMatch[1];
+                if (referencedKey == key)
+                {
+                    std::wstringstream error;
+                    error << L"Self referencing key '" << key << L"' in environment settings";
+                    throw winrt::hresult_error(WEB_E_INVALID_JSON_STRING, winrt::hstring(error.str()));
+                }
+                seenKeys.push_back(key);
+                value.resolvedValue += ResolveEnvironmentVariableValue(referencedKey, envMap, seenKeys);
+                textIter = regexMatch[0].second;
+            }
+            std::copy(textIter, value.rawValue.cend(), std::back_inserter(value.resolvedValue));
+
+            return value.resolvedValue;
+        }
+        const auto size = GetEnvironmentVariableW(key.c_str(), nullptr, 0);
+        if (size > 0)
+        {
+            std::wstring value(static_cast<size_t>(size), L'\0');
+            if (GetEnvironmentVariableW(key.c_str(), value.data(), size) > 0)
+            {
+                // Documentation (https://docs.microsoft.com/en-us/windows/win32/api/processenv/nf-processenv-getenvironmentvariablew)
+                // says: "If the function succeeds, the return value is the number of characters stored in the buffer pointed to by lpBuffer,
+                // not including the terminating null character."
+                // However, my SystemDrive "C:" returns 3 and so without trimming we will end up with a stray NULL string terminator
+                // in the string.
+                value.erase(std::find_if(value.rbegin(), value.rend(), [](wchar_t ch) { return ch != L'\0'; }).base(), value.end());
+                return value;
+            }
+        }

```

Comment:
It might be easier to use the [WIL helper](https://github.com/microsoft/wil/wiki/Win32-helpers) here. I think

```c++
std::wstring value;
auto hr = wil::GetEnvironmentVariableW(key, value);
if (FAILED(hr))
{
    throw winrt::hresult_error(hr);
}
return value;
```

should work

Reply:
That is much nicer, thanks for that!